specs/DisableNavigatingBackAndForward.md

```diff
@@ -0,0 +1,129 @@
+decide whether to suppress. 
+
+This should be solvable in a generic way. However, this feature hasn’t been released yet, and it might
+be better if we could provide a simpler and more direct way. 
+
+Therefore, our job is to provide a mechanism for developers to disable navigating back and forward
+without excessive effort.
+
+
+# Examples
+#### Win32 C++
+
+##### Use `ICoreWebView2NavigationStartingEventArgs3` in `add_NavigationStarting`
+
+```c++
+//! [NavigationStarting]
+// Register a handler for the NavigationStarting event.
+// This handler will check the navigation status, and if the navigation is
+// `GoBack` or `GoForward`, it will be canceled.
+CHECK_FAILURE(m_webView->add_NavigationStarting(
+    Callback<ICoreWebView2NavigationStartingEventHandler>(
+        [this](ICoreWebView2* sender, ICoreWebView2NavigationStartingEventArgs* args)
+        -> HRESULT {
+            wil::com_ptr<ICoreWebView2NavigationStartingEventArgs3> args3;
+            if (SUCCEEDED(args->QueryInterface(IID_PPV_ARGS(&args3))))
+            {
+                COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND history_change =
+                    COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND_OTHER;
+                CHECK_FAILURE(args3->get_NavigationHistoryChange(&history_change));
+                if (history_change != COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND_OTHER)
+                {
+                     CHECK_FAILURE(args->put_Cancel(true));
+                }
+            }
+            return S_OK;
+        })
+    .Get(),
+    &m_navigationStartingToken));
+//! [NavigationStarting]
+```
+
+#### .NET and WinRT
+
+#### Use `CoreWebView2NavigationStartingEventArgs` in `NavigationStarting`
+
+```c#
+// Register a handler for the NavigationStarting event.
+// This handler will check the navigation status, and if the navigation is
+// `GoBack` or `GoForward`, it will be canceled.
+void WebView_NavigationStarting(object sender, CoreWebView2NavigationStartingEventArgs e)
+{
+    if (e.NavigationHistoryChange != CoreWebView2NavigationHistoryChangeKind.Other)
+    {
+        e.Cancel = true;
+    }
+}
+```
+
+# API Details
+#### Win32 C++
+
+```c++
+// Enums and structs
+[v1_enum]
+typedef enum COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND {
+  /// Indicates a navigation that is going back to a previous entry in the navigation history.
+  /// For example, a navigation caused by `CoreWebView2.GoBack` or in script `window.history.go(-1)`.
+  COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND_BACK,
+  /// Indicates a navigation that is going forward to a later entry in the navigation history.
+  /// For example, a navigation caused by `CoreWebView2.GoForward` or in script `window.history.go(1)`.
+  COREWEBVIEW2_NAVIGATION_HISTORY_CHANGE_KIND_FORWARD,
+  /// Indicates a navigation that is not going back or forward to an existing entry in the navigation
+  /// history. For example, a navigation caused by `CoreWebView2.Navigate`, or `CoreWebView2.Reload`
```

Comment:
Is .Reload a kind of navigation we can already detect?  If so, should we add it here?  I can imagine a page doing something like "don't allow reloads" but not caring about blocking Back. Would they use this property to accomplish that?

Reply:
`Reload` can be detected.
```c++
// Returns the offset between the indices of the previous last committed and
// the newly committed navigation entries.
// (e.g. -1 for back navigations, 0 for reloads, 1 for forward navigations).
//
// Note that this value is computed when we create the navigation request
// and doesn't fully cover all corner cases.
// We try to approximate them with params.should_replace_entry, but in
// some cases it's inaccurate:
// - Main frame client redirects,
// - History navigation to the page with subframes. The subframe
//   navigations will return 1 here although they don't create a new
//   navigation entry.
virtual int GetNavigationEntryOffset() = 0;
```

Reply:
What would be the value return by GetNavigationEntryOffset() for direct navigations (not history navigations)?

It seems that GetNavigationEntryOffset is trying to provide a solution to a nonexistent problem. My understanding is that the only thing people are asking for is detecting and possibly blocking forward/back navigation.

Reply:
That was from an original version of the API spec that I think is getting shared again to point out that we have the ability to detect reload and not to suggest this is what the API should look like. The navigation history position index isn't particularly useful because we don't expose the navigation history list and for the stated scenario its easier to use as an enum or Boolean

Reply:
We try to get the `back/forward` status from different source, and the test looks to work.
We get `NavigationType` in `navigation_params.mojom`
```c
enum NavigationType {
  // Reload the page, validating only cache entry for the main resource.
  RELOAD,

  // Reload the page, bypassing any cache entries.
  RELOAD_BYPASSING_CACHE,

  // Reload the page using the original request URL.
  RELOAD_ORIGINAL_REQUEST_URL,

  // The navigation is the result of session restore and should honor the
  // page's cache policy while restoring form state. This is set to true if
  // restoring a tab/session from the previous session and the previous
  // session did not crash. If this is not set and the page was restored then
  // the page's cache policy is ignored and we load from the cache.
  RESTORE,

  // Like RESTORE, except that the navigation contains POST data.
  RESTORE_WITH_POST,

  // History navigation inside the same document.
  HISTORY_SAME_DOCUMENT,

  // History navigation to a different document.
  HISTORY_DIFFERENT_DOCUMENT,

  // Navigation inside the same document. It occurs when the part of the url
  // that is modified is after the '#' part.
  SAME_DOCUMENT,

  // Navigation to another document.
  DIFFERENT_DOCUMENT,
};
```
and classify it:
```
Reload
  RELOAD,
  RELOAD_BYPASSING_CACHE,
  RELOAD_ORIGINAL_REQUEST_URL,

// WebView2 doesn't support session restore so we shouldn't need them.
  RESTORE,
  RESTORE_WITH_POST,

// Navigation inside the same document, can't raise the navigation starting event.
  HISTORY_SAME_DOCUMENT,
  SAME_DOCUMENT,
 
// This should be the back/forward navigation and should get its own enum value. 
BackOrForward
  HISTORY_DIFFERENT_DOCUMENT,

// This should be a navigation to a new document outside of the back/forward navigation and should get its own enum value.
Different
  DIFFERENT,
```